File: Python3/프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    ans = []
    stack = []
    
    msg += " "
    
    # str 사전 초기 셋팅
    str_dict = {}
    last_idx = 1
    for (i, str) in enumerate(range(65,91)):
        str_dict[chr(str)] = i + 1
        last_idx += 1
        
    for m in msg:
        stack_str = "".join(stack)
        new_str = stack_str + m
        if not stack or new_str in str_dict:
            stack.append(m)
        else:
            # 사전추가 작업
            str_dict[new_str] = last_idx
            last_idx += 1
            
            # 출력 작업
            ans.append(str_dict[stack_str])
            stack = []
            
            stack.append(m)
            
    return ans

File: Swift/프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.swift
func solution(_ msg:String) -> [Int] {
    let chr = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    
    var last_idx = 1
    var str_dict = [String: Int]()
    
    for c in chr {
        str_dict[String(c)] = last_idx
        last_idx += 1
    }
    
    var stack: [String] = []
    var result: [Int] = []
    
    for str in msg {
        stack.append(String(str))
        
        let new_str = stack.joined()
        
        if str_dict[new_str] == nil {
            str_dict[new_str] = last_idx
            last_idx += 1
            
            let temp = stack.removeLast()
            result.append(str_dict[stack.joined()]!)
            stack = [temp]
        }
    }
    
    result.append(str_dict[stack.joined()]!)
    return result
}